src/InvoiceManager.php

```diff
@@ -0,0 +1,326 @@
+
+        $product = Product::query()
+                          ->findOrFail($productId);
+        $invoice = $this->getInvoiceById($product->invoice_id);
+        $product->delete();
+
+        return $invoice;
+    }
+
+    public function merge(array $invoiceIds, array $localizedDetails): IInvoice
+    {
+        // TODO: Implement merge() method.
+
+        $first_invoice = $this->getInvoiceById($invoiceIds[0]);
+        $second_invoice = $this->getInvoiceById($invoiceIds[1]);
+
+        return DB::transaction(function () use ($first_invoice, $second_invoice, $localizedDetails) {
+            if (InvoiceStatus::PAID == $first_invoice->getStatus()) {
+                throw new InvalidInvoiceStatusException();
+            }
+            if (InvoiceStatus::PAID == $second_invoice->getStatus()) {
+                throw new InvalidInvoiceStatusException();
+            }
+            if ($first_invoice->getUserId() != $second_invoice->getUserId()) {
+                throw new InvoiceUserMismatchException();
+            }
+            if ($first_invoice->getCurrencyId() !== $second_invoice->getCurrencyId()) {
+                throw new CurrencyMismatchException();
+            }
+
+            return $this->createInvoice($first_invoice, $second_invoice, $localizedDetails);
+        });
+    }
+
+    protected function createInvoice(Invoice $first_invoice, Invoice $second_invoice, array $localizedDetails): IInvoice
+    {
+        $invoice = new Invoice();
+        $invoice->user_id = $first_invoice->getUserId();
+        $invoice->currency_id = $first_invoice->getCurrencyId();
+        if (null != $first_invoice->getMeta() || null != $second_invoice->getMeta()) {
+            $invoice->meta = array_merge($first_invoice->getMeta() ?? [], $second_invoice->getMeta() ?? []);
+        }
+        $invoice->amount = $first_invoice->amount->getValue() + $second_invoice->amount->getValue();
+        if (isset($localizedDetails['title'])) {
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->save();
+        $first_invoice->products()
+                      ->update([
+                                   'invoice_id' => $invoice->getID(),
+                               ]);
+        $second_invoice->products()
+                       ->update([
+                                    'invoice_id' => $invoice->getID(),
+                                ]);
+
+        return $invoice;
+    }
+
+    public function addPaymentToInvoice(int $invoiceId, string $type, INumber $amount, PaymentStatus $status, ?array $meta): IPayment
+    {
+        // TODO: Implement addPaymentToInvoice() method.
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
```

Comment:
Did you check type is valid class name that is implemented `IPaymentMethod` interface?

Reply:
I can't implement it